blockprod/src/detail/block_maker.rs

```diff
@@ -92,20 +93,70 @@ impl BlockMaker {
         Ok(returned_accumulator)
     }
 
-    pub fn make_block(
+    pub async fn make_block(
         &self,
         current_tip_id: Id<Block>,
-        accumulator: &dyn TransactionAccumulator,
+        transactions: &[SignedTransaction],
     ) -> Result<Block, BlockProductionError> {
         // TODO: this isn't efficient. We have to create the header first, then see if it obeys consensus rules, then construct the full block
         let current_time = self.time_getter.get_time();
-        let block = Block::new(
-            accumulator.transactions().clone(),
+        let mut block = Block::new(
+            transactions.to_vec(),
             current_tip_id.into(),
             BlockTimestamp::from_duration_since_epoch(current_time),
             common::chain::block::ConsensusData::None,
             BlockReward::new(vec![]), // TODO: define consensus and rewards through NetworkUpgrades
         )?;
+
+        let consensus_data = self
+            .chainstate_handle
+            .call({
+                let chain_config = Arc::clone(&self.chain_config);
+                let current_tip_height = self.current_tip_height;
+                let header = block.header().clone();
+
+                move |this| {
+                    let get_block_index = |&prev_block_id: &Id<Block>| {
+                        this.get_block_index(&prev_block_id).map_err(|_| {
+                            PropertyQueryError::PrevBlockIndexNotFound(prev_block_id.into())
+                        })
+                    };
+
+                    let get_ancestor = |block_index: &BlockIndex, ancestor_height: BlockHeight| {
+                        this.get_ancestor(
+                            &block_index.clone().into_gen_block_index(),
+                            ancestor_height,
+                        )
+                        .map_err(|_| {
+                            PropertyQueryError::GetAncestorError(
+                                GetAncestorError::InvalidAncestorHeight {
+                                    block_height: block_index.block_height(),
+                                    ancestor_height,
+                                },
+                            )
+                        })
+                    };
+
+                    consensus::generate_consensus_data(
+                        &chain_config,
+                        &header,
+                        current_tip_height,
+                        get_block_index,
+                        get_ancestor,
+                    )
+                }
+            })
+            .await?
+            .map_err(BlockProductionError::FailedConsensusInitialization)?;
+
+        block.update_consensus_data(consensus_data);
+
+        consensus::finalize_consensus_data(
+            &self.chain_config,
+            &mut block,
+            self.current_tip_height,
```

Comment:
It makes more sense to pass the new block height instead of incrementing it inside.

Reply:
Aha, gotcha.